### PR TITLE
Refactor Dimension API and fix a bug

### PR DESCRIPTION
**Pull request by apple502j**
Closes #2474

- Refactors dimension API to store the current target in an entity's field instead of a static field. While memory leak is not a concern, there is a potential issue as described in 2474; however, I'm pretty sure that cannot happen without calling impl code.
- Fixes a bug where end spawn platform gets created when teleporting to the End, if the entity is not a player. This fix also consolidates the 3 mixins into one.
- Add a testmod for teleporting to any dimension.

### CODE REVIEW

**Comment by IMB11** [comment]
> Fixed bug when platform is generated.

Pretty sure that's vanilla

**Comment by apple502j** [comment]
@mineblock11 This is inconsistent with players, which if teleported using this API do not generate the obsidian platform.

**Review by modmuss50 — APPROVED** [review]

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/Teleportable.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.dimension;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.world.TeleportTarget;
+
+public interface Teleportable {
```
Comment (Technici4n) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/Teleportable.java:25]: I'd recommend adding `@ApiStatus.Internal`.

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/FabricDimensionInternals.java [hunk 2]
```diff
@@ -23,32 +23,18 @@
 import net.minecraft.server.world.ServerWorld;
 import net.minecraft.world.TeleportTarget;
 
-import net.fabricmc.fabric.api.dimension.v1.FabricDimensions;
-
 public final class FabricDimensionInternals {
```
Comment (Technici4n) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/FabricDimensionInternals.java:28]: Could also have `@ApiStatus.Internal`.

**Review by Technici4n — APPROVED** [review]
Good on principle.

**Pull request merged**